tokens/src/commands.rs

```diff
@@ -261,16 +290,7 @@ async fn distribute_allocations(
             }
         };
     }
-    if args.dry_run {
-        let undistributed_tokens: f64 = allocations.iter().map(|x| x.amount).sum();
-        check_payer_balances(
-            num_signatures,
-            sol_to_lamports(undistributed_tokens),
-            client,
-            args,
-        )
-        .await?;
-    }
+    check_payer_balances(num_signatures, allocations, client, args).await?;
```

Comment:
@garious there's an outstanding issue here, and I'd appreciate your thoughts. Now that this check happens on real-deal runs, it needs to move, otherwise the first allocation gets attempted and logged in the db. It probably makes more sense to for this check to be outside `distribute_allocations` anyway.

Can you recommend a neat way to determine the number of signatures for the distribution? (ideally without duplicating all of this https://github.com/solana-labs/solana/pull/12380/files#diff-f5393ff0891e002c6d8f29fe8b90fb07R235-L223)

Reply:
I don't think I understand the problem. As I recall, everything logged to the db during a dry run is discarded.

Reply:
Yep, but this PR promotes `check_payer_balances` to run on non-dry runs as well.

Reply:
I'm still a little lost, but if you refactor such that you collect all the Messages upfront, then you can calculate `num_signatures` using pub field `message.header.num_required_signatures`.  The only important aspect of the ordering here is that the Transaction is sent immediately after getting the blockhash and signing so that the blockhash won't expire.